### PR TITLE
Supporting bind candidates

### PR DESCRIPTION
**Pull request by Shitaibin**
Candidates and binder can make a deal to make binder stake 1000VNT for candidate. The binder can assign an account to get the reward including producing reward and voting reward of the candidate. And the candidate do not receive the reward any more.

The only registered and bind candidate has the opportunity to become to witness.

The binder can be the candidate itself. But it's not recommend for safety.

### CODE REVIEW

File: core/vm/election/abi.go [hunk 1]
```diff
@@ -23,7 +23,7 @@ import (
 )
 
 const ElectionAbiJSON = `[
-{"name":"registerWitness","inputs":[{"name":"nodeUrl","type":"bytes"},{"name":"website","type":"bytes"},{"name":"nodeName","type":"bytes"}],"outputs":[],"type":"function"},
+{"name":"registerWitness","inputs":[{"name":"nodeUrl","type":"bytes"},{"name":"website","type":"bytes"},{"name":"nodeName","type":"bytes"},{"name":"locker","type":"address"},{"beneficiary":"nodeName","type":"address"}],"outputs":[],"type":"function"},
```
Comment (Shitaibin) [inline core/vm/election/abi.go]: 1. locker -> binder
2. -> "name":"beneficiary"

File: core/vm/election/election.go [hunk 2]
```diff
@@ -200,6 +220,19 @@ func (e *Election) RequiredGas(input []byte) uint64 {
...
+	NodeUrl     []byte
+	Website     []byte
+	NodeName    []byte
+	Locker      common.Address // 绑定人
+	Beneficiary common.Address // 受益人
+}
+
+type BindInfo struct {
+	Candidate   common.Address // 绑定的候选账号
+	beneficiary common.Address // 收益的账号
```
Comment (Shitaibin) [inline core/vm/election/election.go]: 重构：改为大写

File: core/vm/election/election.go [hunk 3]
```diff
@@ -390,9 +431,90 @@ func (ec electionContext) unregisterWitness(address common.Address) error {
...
+	}
+	return nil
+}
+
+// bindCandidate 绑定候选节点，绑定人受益人信息需与候选人注册信息一致
+func (ec electionContext) bindCandidate(locker common.Address, info *BindInfo, amount *big.Int) error {
+	candi := info.Candidate
+	beneficiary := info.beneficiary
+
+	// Check lock amount
```
Comment (Shitaibin) [inline core/vm/election/election.go]: bind amount

File: core/vm/election/election.go [hunk 4]
```diff
@@ -390,9 +431,90 @@ func (ec electionContext) unregisterWitness(address common.Address) error {
...
+
+	// 返回绑定人锁仓金额
+	return ec.transfer(contractAddr, locker, bindAmount)
+}
+
+func (ec electionContext) matchLockerAndCandi(locker, candi, beneficiary common.Address) (*Candidate, error) {
+	// get candidate from db
+	candidate := ec.getCandidate(candi)
+
+	// if candidate is active
```
Comment (Shitaibin) [inline core/vm/election/election.go]: 注释删除

File: core/vm/election/electiondb_test.go [hunk 5]
```diff
@@ -94,8 +97,14 @@ func sameVoter(voter *Voter, voter1 *Voter) (bool, error) {
 func sameCandidate(candidate *Candidate, candidate1 *Candidate) (bool, error) {
 	if !bytes.Equal(candidate.Owner[:], candidate1.Owner[:]) {
 		return false, fmt.Errorf("Error,owner before %v and after %v is different", candidate.Owner, candidate1.Owner)
-	} else if candidate.Active != candidate1.Active {
-		return false, fmt.Errorf("Error,active before %v and after %v is different", candidate.Active, candidate1.Active)
+	} else if candidate.Registered != candidate1.Registered {
+		return false, fmt.Errorf("Error,registered before %v and after %v is different", candidate.Registered, candidate1.Registered)
+	} else if candidate.Binder != candidate1.Binder {
+		// 	TODO
```
Comment (Shitaibin) [inline core/vm/election/electiondb_test.go]: 补充

File: core/vm/election/electiondb_test.go [hunk 6]
```diff
@@ -472,7 +484,7 @@ func TestGetFirstXCandidates_3(t *testing.T) {
 		candidate1 := candidate
 		candidate1.Owner[0] = byte(tests[i].addrPre)
 		candidate1.VoteCount = big.NewInt(tests[i].votes)
-		candidate1.Active = tests[i].active
+		candidate1.Registered = tests[i].active
```
Comment (Shitaibin) [inline core/vm/election/electiondb_test.go]: 修改用例

File: core/vm/election/electiondb_test.go [hunk 7]
```diff
@@ -534,7 +546,7 @@ func TestGetFirstXCandidates_4(t *testing.T) {
 		candidate1 := candidate
 		candidate1.Owner[0] = byte(tests[i].addrPre)
 		candidate1.VoteCount = big.NewInt(tests[i].votes)
-		candidate1.Active = true
+		candidate1.Registered = true
```
Comment (Shitaibin) [inline core/vm/election/electiondb_test.go:539]: 修改用例

File: core/vm/election/election.go [hunk 8]
```diff
@@ -390,9 +415,91 @@ func (ec electionContext) unregisterWitness(address common.Address) error {
...
+	if !candidate.Registered {
+		return ErrCandiNotReg
+	}
+	if !candidate.Bind {
+		return ErrCandiNotBind
+	}
+
+	// 取消绑定
+	candidate.Bind = false
+	candidate.Binder = emptyAddress
```
Comment (Shitaibin) [inline core/vm/election/election.go]: 删除

**Review by ruopuzi — APPROVED** [review]

**Review by ooozws — APPROVED** [review]

**Pull request merged**